### PR TITLE
Streamline conversion to/from IValue

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #21317 Remove double indirection for storing lists and dicts&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15615077/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#21228 Streamline conversion to/from IValue**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15585806/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #21177 Use c10::List&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15476433/)

Before, we had IValue::to<> and kernel_functor.h had its own ivalue_to_arg_type and return_type_to_ivalue.
They are now unified.

Also, this means that nested types like Dicts of Lists of Optional of Dict of ... do work as expected now.

Differential Revision: [D15585806](https://our.internmc.facebook.com/intern/diff/D15585806/)

### CODE REVIEW

File: aten/src/ATen/core/ivalue_inl.h [hunk 1]
```diff
@@ -623,31 +660,54 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::Future> v)
...
 }
 
-inline const c10::ListPtr<int64_t>& IValue::toIntListRef() const {
+inline const c10::ListPtr<int64_t>& IValue::toIntListRef() const & {
   return toIntList()->elements();
 }
 
-inline const c10::ListPtr<double>& IValue::toDoubleListRef() const {
+inline c10::ListPtr<int64_t> IValue::toIntListRef() && {
+  return std::move(*this).toIntList()->elements();
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue_inl.h]: I am really confused by all the different list type here.

There is at least:
1. std::vector<T>, the standard C++ list.
2. std::intrusive_ptr<List<int|tensor|double|bool|IValue>, what is held in an ivalue
3. c10::ListPtr<int64_t>. Is this a wrapper around 2? Or does 2 hold 3?, Does that add an unnecessary indirection?

**Review by zdevito — APPROVED** [review]
Changes specific to this patch seem good. I still do not think the c10::ListPtr patch is ready because it is adding complexity all over ivalue.h's implementation.

File: aten/src/ATen/core/ivalue_inl.h (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/core/ivalue_inl.h]: std::vector and c10::ListPtr will stay. I'm currently working on another diff that would streamline this, remove the double indirection IValue->ivalue::List->c10::ListPtr in favor of directly storing c10::ListPtr in IValue, and if possible, remove the intrusive_ptr versions from the interface. That is unrelated to this PR though.

**Comment by smessmer** [comment]
merged into https://github.com/pytorch/pytorch/pull/21177

**Pull request closed without merging**